[JSC] Cache lookups of ProxyObject's traps

shvaikalesh opened the pull request:
#### e1e2cbbf2d6f36f82c6b7094312c230f6d152a2a
<pre>
[JSC] Cache lookups of ProxyObject&apos;s traps
<a href="https://bugs.webkit.org/show_bug.cgi?id=256554">https://bugs.webkit.org/show_bug.cgi?id=256554</a>
&lt;rdar://problem/109119378&gt;

Reviewed by Yusuke Suzuki.

This patch introduces a caching of handler&apos;s traps by storing an array of property offsets, handler&apos;s
structure ID, and handler&apos;s prototype structure ID. We optimize for common case of ProxyObject&apos;s handler
being a plain JSFinalObject inheriting from Object.prototype, which [[Prototype]] is always `null`.

For now, only 3 of the most popular traps are cached: &quot;has&quot;, &quot;get&quot;, and &quot;set&quot;.
Also, removes extra isCell() check from getMethod() to micro-optimize lookup of other traps.

This change progresses microbenchmarks with missing handlers by 60-70% when IC is off,
and speeds up Speedometer2/Flight-TodoMVC by 2%.

                                   ToT                      patch

proxy-has-miss-handler      310.7297+-1.4310     ^    182.3500+-0.2509        ^ definitely 1.7040x faster
proxy-set-miss-handler     1305.7705+-2.9307     ^   1179.7164+-4.0497        ^ definitely 1.1069x faster
proxy-get-miss-handler      310.6261+-0.1537     ^    190.5240+-0.3619        ^ definitely 1.6304x faster

&lt;geometric&gt;                 501.3743+-1.0979     ^    344.7815+-0.4166        ^ definitely 1.4542x faster

* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::getMethod):
* Source/JavaScriptCore/runtime/ProxyObject.cpp:
(JSC::ProxyObject::ProxyObject):
(JSC::ProxyObject::getHandlerTrap):
(JSC::ProxyObject::clearHandlerTrapsOffsetsCache):
(JSC::performProxyGet):
(JSC::ProxyObject::performHasProperty):
(JSC::ProxyObject::performPut):
(JSC::ProxyObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/ProxyObject.h:

Canonical link: <a href="https://commits.webkit.org/263944@main">https://commits.webkit.org/263944@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/117834751c64c21c67cfcc3d38e87aa624803b4e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6153 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6340 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6522 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7715 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6492 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6560 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6292 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9372 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6263 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6280 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5572 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7780 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3754 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5552 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13450 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/5169 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5619 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5631 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7863 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/5747 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6097 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4991 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/6274 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5515 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1494 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1468 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9658 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/6449 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5882 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1620 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/ProxyObject.cpp:
```diff
@@ -82,6 +86,64 @@ void ProxyObject::finishCreation(VM& vm, JSGlobalObject* globalObject, JSValue t
+
+    bool isSlotCacheable = slot.isUnset() || (slot.isCacheableValue() && slot.slotBase() == handler);
+    if (isSlotCacheable) {
+        bool isHandlerPrototypeChainCacheable = handler->type() == FinalObjectType && !handler->structure()->isDictionary() && handler->getPrototypeDirect() == objectPrototype && !objectPrototype->structure()->isDictionary();
```
Need to check `JSGlobalObject::objectPrototypeChainIsSane` too, and add test changing `Object.prototype.__proto__`.

Constellation replied:
Cacheability depends on unset case / set case.
1. If this is set case, then we can cache offset even if `handler->getPrototypeDirect() == objectPrototype` is not met because this is own property case. And cacheable dictionary works since it is own hitting property. We should check uncacheable dictionary in this case instead.
2. If this is unset case, then we need to check dictionary condition, and prototype chains.

Constellation replied:
But since we would like to reflect all three fields with this one cache, the above different would not be necessary. But anyway, we need to check `objectPrototypeChainIsSane`

shvaikalesh replied:
Discussed `objectPrototypeChainIsSane` thing offline with Yusuke: it's not actually necessary since `Object.prototype` is an immutable prototype object.

shvaikalesh replied:
As for cacheability condition, we can certainly make it more sophisticated, but even the current one covers 99.9% cases so why bother.

Constellation commented on Source/JavaScriptCore/runtime/ProxyObject.cpp:
```diff
@@ -82,6 +86,64 @@ void ProxyObject::finishCreation(VM& vm, JSGlobalObject* globalObject, JSValue t
+        return asObject(value);
+    };
+
+    JSObject* objectPrototype = globalObject->objectPrototype();
```
This `globalObject` is lexical one, but we should check against ProxyObject's JSGlobalObject.

shvaikalesh replied:
YES thank you that was a real bug right there, and even switching to ProxyObject's wouldn't be a 100% correct fix. We can't avoid calling `handler->getPrototypeDirect()`.

Constellation commented on Source/JavaScriptCore/runtime/ProxyObject.cpp:
```diff
@@ -82,6 +86,64 @@ void ProxyObject::finishCreation(VM& vm, JSGlobalObject* globalObject, JSValue t
+        if (offset != emptyHandlerTrapCache)
+            return ensureIsCallable(handler->getDirect(offset));
+    } else
+        clearHandlerTrapsOffsetsCache();
```
We should do this only when `m_handlerStructureID` is set.

shvaikalesh replied:
Done!

Constellation commented on Source/JavaScriptCore/runtime/ProxyObject.h:
```diff
@@ -129,6 +141,9 @@ class ProxyObject final : public JSInternalFieldObjectImpl<2> {
 
     bool m_isCallable : 1 { false };
     bool m_isConstructible : 1 { false };
+    PropertyOffset m_handlerTrapsOffsetsCache[3];
```
Use `numberOfCachedHandlerTrapsOffsets`.

shvaikalesh replied:
Done!

Constellation commented on Source/JavaScriptCore/runtime/ProxyObject.cpp:
```diff
@@ -82,6 +86,64 @@ void ProxyObject::finishCreation(VM& vm, JSGlobalObject* globalObject, JSValue t
+    if (isSlotCacheable) {
+        bool isHandlerPrototypeChainCacheable = handler->type() == FinalObjectType && !handler->structure()->isDictionary() && handler->getPrototypeDirect() == objectPrototype && !objectPrototype->structure()->isDictionary();
+        if (isHandlerPrototypeChainCacheable) {
+            m_handlerTrapsOffsetsCache[static_cast<uint8_t>(trap)] = slot.cachedOffset();
```
To make it defensive, let's avoid caching the offset if the offset is `emptyHandlerTrapCache`

shvaikalesh replied:
I've made `emptyHandlerTrapCache` an `INT_MIN` and added an assert that cached offset never equals it. Like we validate that all offsets are valid ie `> -1`.

Constellation commented on Source/JavaScriptCore/runtime/ProxyObject.cpp:
```diff
@@ -82,6 +86,65 @@ void ProxyObject::finishCreation(VM& vm, JSGlobalObject* globalObject, JSValue t
+        if (isHandlerPrototypeChainCacheable) {
+            ASSERT(slot.cachedOffset() != emptyHandlerTrapCache);
+            m_handlerTrapsOffsetsCache[static_cast<uint8_t>(trap)] = slot.cachedOffset();
+            m_handlerStructureID.set(vm, this, handler->structure());
+            m_handlerPrototypeStructureID.set(vm, this, asObject(handlerPrototype)->structure());
```
We need to check that already-cached offsets are still valid with this structures, right?
What happens if,

1. Structure A and B, then cache the offset for Get
2. Structure changed to C and D
3. Then, cache C and D and cache the offset for Has
4. But still Get is cached, and it should be invalidated.

shvaikalesh replied:
Discussed offline, once any of the structures is change we clear cache for all the traps, ending up with cache for only `Has` trap that is correct:

```cpp
    bool isCacheValid = handler->structureID() == m_handlerStructureID.value() && asObject(handlerPrototype)->structureID() == m_handlerPrototypeStructureID.value();
    if (isCacheValid) {
       ...
    } else if (m_handlerStructureID)
        clearHandlerTrapsOffsetsCache();
```

Constellation reviewed (approved):
r=me

Commit:
[JSC] Cache lookups of ProxyObject's traps
https://bugs.webkit.org/show_bug.cgi?id=256554
<rdar://problem/109119378>

Reviewed by Yusuke Suzuki.

This patch introduces a caching of handler's traps by storing an array of property offsets, handler's
structure ID, and handler's prototype structure ID. We optimize for common case of ProxyObject's handler
being a plain JSFinalObject inheriting from Object.prototype, which [[Prototype]] is always `null`.

For now, only 3 of the most popular traps are cached: "has", "get", and "set".
Also, removes extra isCell() check from getMethod() to micro-optimize lookup of other traps.

This change progresses microbenchmarks with missing handlers by 60-70% when IC is off,
and speeds up Speedometer2/Flight-TodoMVC by 2%.

                                   ToT                      patch

proxy-has-miss-handler      310.7297+-1.4310     ^    182.3500+-0.2509        ^ definitely 1.7040x faster
proxy-set-miss-handler     1305.7705+-2.9307     ^   1179.7164+-4.0497        ^ definitely 1.1069x faster
proxy-get-miss-handler      310.6261+-0.1537     ^    190.5240+-0.3619        ^ definitely 1.6304x faster

<geometric>                 501.3743+-1.0979     ^    344.7815+-0.4166        ^ definitely 1.4542x faster

* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::getMethod):
* Source/JavaScriptCore/runtime/ProxyObject.cpp:
(JSC::ProxyObject::ProxyObject):
(JSC::ProxyObject::getHandlerTrap):
(JSC::ProxyObject::clearHandlerTrapsOffsetsCache):
(JSC::performProxyGet):
(JSC::ProxyObject::performHasProperty):
(JSC::ProxyObject::performPut):
(JSC::ProxyObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/ProxyObject.h:

Canonical link: https://commits.webkit.org/263944@main
